File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        
        for i in asteroids:
            while stack and i < 0 < stack[-1]:
                if stack[-1] < -i:
                    stack.pop()
                    continue  # Continue checking the next element in the stack
                elif stack[-1] == -i:
                    stack.pop()
                    break  # Both are destroyed, exit the loop
                break  # Current asteroid is destroyed, exit the loop
            else:
                stack.append(i)  # Append the asteroid if no collision occurs
                
        return stack

File: 735-asteroid-collision/asteroid-collision.js
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
    let stack = [];

    for (let a of asteroids) {
        let destroyed = false;

        while (
            stack.length &&
            stack[stack.length - 1] > 0 &&
            a < 0
        ) {
            let top = stack[stack.length - 1];
            if (top < -a) {
                stack.pop(); // top explodes
            } else if (top === -a) {
                stack.pop(); // both explode
                destroyed = true;
                break;
            } else {
                // current asteroid explodes
                destroyed = true;
                break;
            }
        }

        if (!destroyed) {
            stack.push(a);
        }
    }

    return stack;
};